### PR TITLE
Additional Build Documentation added.

### PR DESCRIPTION
**Pull request by SayantanMajumdar**
In response to #788 added file `docs/Additional_Build_Documentation.md` containing
additional build documentation which attempts to accieve the following:
   - demonstarte the use of BENCHMARK_DOWNLOAD_DEPENDENCIES
   - vcpkg/installation
   - Building from  head 
   - embedding Benchmark in other CMake projects

Then added a link to the file `docs/Additional_Build_Documentation.md` in `README.md`.

### CODE REVIEW

File: docs/Additional_Build_Documentation.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+```
+To configure Benchmark we use CMake as follows:
+```bash
+$ cmake -G <generator> [options] -S benchmark -B build
+```
+If we haven't cloned the googletest repository as described above we can ask benchmark to download this dependency in the above configuration step by adding `-DBENCHMARK_DOWNLOAD_DEPENDENCIES=ON`
+
+Then to build Benchmark we use CMake again as follows:
+```bash
+$ cmake -build build
```
Comment (LebedevRI) [inline docs/Additional_Build_Documentation.md]: `--build` ?

File: docs/Additional_Build_Documentation.md [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+### In Windows
+```bash
+> cl /EHsc /MD BenchmarkEverything.cpp /I<BENCHMARK_SOURCE_DIR>\include /link shlwapi.lib <BENCHMARK_BUILD_DIR>\src\Release\benchmark.lib
+```
+where `<BENCHMARK_SOURCE_DIR>` is the directory where Benchmark was cloned into and `<BENCHMARK_BUILD_DIR>` the directory where Benchmark was build. Here, we had build Benchmark in the Release configuration and hence we 
+used `<BENCHMARK_BUILD_DIR>\src\Release\benchmark.lib`. We also used the `cl` compiler which is provided with Visual Studio and also linked the `shlwapi` library.
+
+### In Linux
+```bash
+$ g++ -std=c++11 BenchmarkEverything.cpp -I<BENCHMARK_SOURCE_DIR>/include -pthread -L <BENCHMARK_BUILD_DIR>\src -lbenchmark
```
Comment (LebedevRI) [inline docs/Additional_Build_Documentation.md]: wrong slash

File: docs/Additional_Build_Documentation.md [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+```
+where `<BENCHMARK_SOURCE_DIR>` is the directory where Benchmark was cloned into and `<BENCHMARK_BUILD_DIR>` the directory where Benchmark was build. We also need to link the `pthread` library.
+
+# Installing Benchmark
+
+After building Benchmark we can choose to install it in our systems. 
+
+For Ubuntu and Debian based systems we have to build Benchmark in the Release configuration by applying `-DCMAKE_BUILD_TYPE=RELEASE` in the configuration step. Then install it using `sudo make install` inside the build directory.
```
Comment (LebedevRI) [inline docs/Additional_Build_Documentation.md]: please linewrap everything to 80 col

File: docs/Additional_Build_Documentation.md [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+Now, Benchmark uses Google Test to run it's tests. So, if we wish to build and run
+the tests we need to include googletest, which can be done either by cloning the googletest
+repository or by asking benchmark in the configuration step to download this dependency.
+We can clone the googletest repository in benchmark using
+```bash
+$ git clone https://github.com/google/googletest.git benchmark/googletest
+```
+To configure Benchmark we use CMake as follows:
+```bash
+$ cmake -G <generator> [options] -S benchmark -B build
```
Comment (dominichamon) [inline docs/Additional_Build_Documentation.md]: you should differentiate between the cmake and bazel steps.

File: docs/Additional_Build_Documentation.md (hunk 1 above)
Comment (dominichamon) [inline docs/Additional_Build_Documentation.md]: alternatively, users can use cmake to generate and then Visual Studio or whatever build system they want to build.

File: docs/Additional_Build_Documentation.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+```
+If we haven't cloned the googletest repository as described above we can ask benchmark to download this dependency in the above configuration step by adding `-DBENCHMARK_DOWNLOAD_DEPENDENCIES=ON`
```
Comment (dominichamon) [inline docs/Additional_Build_Documentation.md]: this is only if you want to build from head. if you just want to _use_ benchmark then we should document that separately.

File: docs/Additional_Build_Documentation.md [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+If we don't want the Benchmark tests to build and run we can add `-DBENCHMARK_ENABLE_GTEST_TESTS=OFF` in the configuration step.
+
+## Using Benchmark
+
+If we want to use Benchmark with a particular file, say `BenchmarkEverything.cpp`,
+we can build Benchmark and use the following comands in the terminal.
+
+### In Windows
+```bash
+> cl /EHsc /MD BenchmarkEverything.cpp /I<BENCHMARK_SOURCE_DIR>\include /link shlwapi.lib <BENCHMARK_BUILD_DIR>\src\Release\benchmark.lib
```
Comment (dominichamon) [inline docs/Additional_Build_Documentation.md]: i presume there's an easier way if you're using VS

File: docs/Additional_Build_Documentation.md (hunk 2 above)
Comment (dominichamon) [inline docs/Additional_Build_Documentation.md]: if you're going to prescribe a compiler, consider clang++ :)

File: README.md [hunk 7]
```diff
@@ -40,6 +40,7 @@ dependency can be provided two ways:
 If you do not wish to build and run the tests, add `-DBENCHMARK_ENABLE_GTEST_TESTS=OFF`
 to `CMAKE_ARGS`.
 
+For a more detailed documentation of building benchmark checkout [Additional Build Documentation](docs/Additional_Build_Documentation.md)
```
Comment (dominichamon) [inline README.md]: this also covers USING benchmark.

Perhaps the docs would be clearer if we separated these two use-cases. One doc for building and installing from source, one doc for using it (linking to building from source, but also having instructions for `apt` or `vcpkg` or whatever).

File: docs/Additional_Build_Documentation.md [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+For Windows
+```bash
+PS> .\vcpkg install benchmark
+```
+For Linux
+```bash
+$ ./vcpkg install benchmark
+```
+
+# Embedding Benchmark in other projects
```
Comment (dominichamon) [inline docs/Additional_Build_Documentation.md]: as above, this should come first _or_ be a different doc.

File: docs/Additional_Build_Documentation.md (hunk 3 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: what about the commands? do they have to be under 80 col?

File: docs/Additional_Build_Documentation.md (hunk 1 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: yeah, will add that. I feared that might create a situation where we have to give instructions for each generator. but will add it.

File: docs/Additional_Build_Documentation.md (hunk 6 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: yes, others will have syntax similar to linux(I think)

File: docs/Additional_Build_Documentation.md (hunk 1 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: will fix it

File: docs/Additional_Build_Documentation.md (hunk 2 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: will fix it

**Comment by SayantanMajumdar** [comment]
Moved the installation documentation for `vcpkg` into the `README.md` and added a hint for the VS 15 2017 generator in the build instructions in `README.md`. I removed the build instructions in my docs file  `docs/Additional_Build_Documentation.md` as it was redundant and it was well written in the `README.md`.

Renamed the file `docs/Additional_Build_Documentation.md` to `docs/Using_Benchmark.md` with instructions of using Benchmark and embedding it in other projects. 
Instructions on 
-    Using Benchmark with single file projects in the terminal
-    Including Benchmark as a submodule in a project.
-    Making CMake download and build Benchmark as a part of the project.

**Comment by SayantanMajumdar** [comment]
Moved the installation instructions for `vcpkg` to `README.md` from `docs/Additional_Build_Documentation.md` and added a simple example for building Benchmark using the VS 15 2017 for VS users. The build instructions in `docs/Additional_Build_Documentation.md` were redundent and it was already well written in `README.md` .

Renamed `docs/Additional_Build_Documentation.md` to `docs/Using_Benchmark.md` and included documentation for using Benchmark with small single file projects, documentation about using it as submodule in other projects and making CMake download and build Benchmark as a part of the project just like googletest

**Comment by SayantanMajumdar** [comment]
Moved the installation instructions for `vcpkg` to `README.md` from `docs/Additional_Build_Documentation.md` and added a simple example for building Benchmark using the VS 15 2017 for VS users. The build instructions in `docs/Additional_Build_Documentation.md` were redundent and it was already well written in `README.md` .

Renamed `docs/Additional_Build_Documentation.md` to `docs/Using_Benchmark.md` and included documentation for using Benchmark with small single file projects, documentation about using it as submodule in other projects and making CMake download and build Benchmark as a part of the project just like googletest

File: README.md (hunk 7 above)
Comment (SayantanMajumdar) [reply README.md]: made a separate file for embedding Benchmark and moved the `vcpkg` installation to `README.md`

File: docs/Additional_Build_Documentation.md (hunk 8 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: embedding and using is in the file `docs/Using_Benchmark.md` and the installation using `vcpkg` has been moved to `README.md`

File: docs/Additional_Build_Documentation.md (hunk 2 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: done

File: docs/Additional_Build_Documentation.md (hunk 5 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: using has been put in a separate file now

File: docs/Additional_Build_Documentation.md (hunk 1 above)
Comment (SayantanMajumdar) [reply docs/Additional_Build_Documentation.md]: added example for VS 15 2017, the build documentation in `README.md` was good and so added the example there. and removed the build documentation form the file added by me.

File: README.md [hunk 9]
```diff
@@ -29,6 +29,11 @@ $ cmake -G <generator> [options] ../benchmark
 $ make
 ```
 
+Similarly for the Visual Studio 15 2017 generator we can use
+```bash
+> msbuild benchmark.sln /p:Configuration=Release /p:Platform=Win32
```
Comment (LebedevRI) [inline README.md]: Do we want to spell it out for every cmake generator?
Maybe let's just
```diff
-$ make
+$ cmake --build ./
```
that is more correct than hardcoding `make` there anyway

File: README.md [hunk 10]
```diff
@@ -70,6 +75,26 @@ If you need to install the library globally
...
+(NOTE: documentation for installation of `vcpkg` can be found 
+[here](https://github.com/Microsoft/vcpkg))
+
+For Windows
+```bash
+PS> .\vcpkg install benchmark
+```
+For Linux
+```bash
+$ ./vcpkg install benchmark
```
Comment (LebedevRI) [inline README.md]: These expect that `vcpkg` is located in current directory, and only differ in slashes.
Can this be condensed into a single entry?

File: docs/Using_Benchmark.md [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+> cl /EHsc /MD BenchmarkEverything.cpp /I<BENCHMARK_SOURCE_DIR>\include /link shlwapi.lib <BENCHMARK_BUILD_DIR>\src\Release\benchmark.lib
+```
+where, `<BENCHMARK_SOURCE_DIR>` is the directory where Benchmark was cloned into
+and `<BENCHMARK_BUILD_DIR>` the directory where Benchmark was built. Here, we 
+had built Benchmark in the Release configuration and hence we used 
+`<BENCHMARK_BUILD_DIR>\src\Release\benchmark.lib`. We also used the `cl` 
+compiler which is provided with Visual Studio and also linked the `shlwapi` 
+library.
+
+## In Linux
```
Comment (LebedevRI) [inline docs/Using_Benchmark.md]: This is super flame-wary and nitpicky, but i think linux should be explained first.

File: docs/Using_Benchmark.md [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+set(BENCHMARK_ENABLE_TESTING OFF CACHE BOOL "Disable benchmark tests" FORCE)
+add_subdirectory(extprojects/benchmark)
+# including benchmark header files
+include_directories(extprojects/benchmark/include)
+add_executable(BenchmarkEverything BenchmarkEverything.cpp)
+target_link_libraries(BenchmarkEverything benchmark)
+if(WIN32)
+    target_link_libraries(BenchmarkEverything shlwapi)
+elseif(LINUX)
+    target_link_libraries(BenchmarkEverything pthread)
```
Comment (LebedevRI) [inline docs/Using_Benchmark.md]: This is a bad advise.
The library needs to be found first.

File: docs/Using_Benchmark.md [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+$ cmake -S . -B <BUILD_DIR> -G <GENERATOR>
+$ cmake --build <BUILD_DIR> --config <CONFIGURATION>
+```
+where `<BUILD_DIR>` is the directory where you would like to build 
+`BenchmarkEverything`, `<GENERATOR>` is the generator that you would 
+like to use and `<CONFIGURATION>` the configuration that you would like 
+to build, like `Release` or `Debug`. To run `BenchmarkEverything` we just 
+need to run the executable generated inside `<BUILD_DIR>`.
+
+## Use CMake for embedding Benchmark
```
Comment (LebedevRI) [inline docs/Using_Benchmark.md]: There is one more variant - download it at cmake time, and then fall-back into `Adding Benchmark as a git submodule` variant.

File: docs/Using_Benchmark.md (hunk 11 above)
Comment (SayantanMajumdar) [reply docs/Using_Benchmark.md]: its the same

File: docs/Using_Benchmark.md (hunk 13 above)
Comment (SayantanMajumdar) [reply docs/Using_Benchmark.md]: that would require you understand both given options first. someone doing that probably is an advanced coder

File: README.md (hunk 10 above)
Comment (SayantanMajumdar) [reply README.md]: will fix it

File: docs/Using_Benchmark.md (hunk 12 above)
Comment (SayantanMajumdar) [reply docs/Using_Benchmark.md]: I tried to keep it simple if the library can't be found the error will say that

Comment (LebedevRI) [reply docs/Using_Benchmark.md]: link-time error
link-time errors are not nice :/

**Pull request closed without merging**